### ISSUE TITLE
Idea: lint when passing async functions into forEach

### ISSUE THREAD

**Issue by greglittlefield-wf** [dart-lang/linter#891]
I recently ran into a few instances of code that used the following:
```dart
values.forEach((value) async {
  // ...
  doSomethingSync();
  await value.doSomethingAsync();
});
```
```dart
map.forEach((key, value) async {
  doSomethingSync();
  await value.doSomethingAsync();
});
```

The intended behavior in at least one of the cases was to wait for all of the `await`s to be run before proceeding, which can be done via `Future.forEach` or a for loop with `awaits`. 

Even if the intended behavior is what is actually happening, it might not be immediately obvious to someone unfamiliar with how `async` functions are executed in Dart 1.x.

The first example would be avoided by using the [prefer_foreach](http://dart-lang.github.io/linter/lints/prefer_foreach.html) rule (unless a tearoff were passed in instead), but the second would not. Still, for the first example, it might be good to have a more specific lint to catch this case.

Are `Iterable.forEach`/`Map.forEach` common-enough special cases where using async functions should result in a lint? 

Another idea, though it would have far more impact and may be too noisy, would be to lint when an async function is being used where a `void` function is accepted.
```dart
CallbackSubscription registerVoidCallback(void callback()) { /*...*/ }

// Good
registerVoidCallback((() { /* ... */ });
// Would lint
registerVoidCallback((() async { /* ... */ });
```

**Comment by matanlurey** [comment]
Async functions will be executed synchronously in Dart 2, FWIW.

**Comment by greglittlefield-wf** [comment]
Good point! This lint would definitely have more impact in Dart 1.

**Comment by natebosch** [comment]
See also https://github.com/dart-lang/linter/issues/1123

**Comment by jamesderlin** [comment]
I see this confusion *all the time* on StackOverflow.  It would be nice if there were a warning about passing `Future`-returning callbacks to `Iterable.forEach`/`Map.forEach`.

**Comment by pq** [comment]
@jamesderlin: the next time you see this come up  on SO, could you add a link?  That'd be great for motivation.  Thanks!

**Comment by jamesderlin** [comment]
Here are some of them:
* [Async/await in List.forEach()](https://stackoverflow.com/q/42467663/)
* [[dart]flutter how to wait map.forEach to complete in Future<void> function?](https://stackoverflow.com/q/63719374/)
* [Waiting async in Flutter](https://stackoverflow.com/q/61224843/)
* [list.forEach doesn't await properly in async method](https://stackoverflow.com/q/57324877/)
* [Strange async behaviour](https://stackoverflow.com/q/31535915/)
* [Flutter Await not waiting](https://stackoverflow.com/q/62734766/)
* [Flutter - Dart : wait a forEach ends](https://stackoverflow.com/questions/53886373/)
* [My async call is returning before list is populated in forEach loop](https://stackoverflow.com/q/51106934/)
* [How to avoid using await key in dart Map by foreach function](https://stackoverflow.com/q/52276412/)
* [Flutter async/await doesn't work inside forEach](https://stackoverflow.com/q/63466257/)
* [Flutter await not waiting before continuing](https://stackoverflow.com/q/60084469/)

**Comment by pq** [comment]
Amazing.  Thank you!
